### PR TITLE
Productionise webview ≥ 77 check

### LINKED ISSUE #15855: Productionise WebView >= 77 check
It'd be worth having the dev-only check introduced in https://github.com/ankidroid/Anki-Android/issues/14702 also explain to regular users how to upgrade.

Android 6 supports >= WebView 77, so a user should be able to upgrade

* https://github.com/ankidroid/Anki-Android/issues/14702

_Originally posted by @david-allison in https://github.com/ankidroid/Anki-Android/issues/15847#issuecomment-1988370221_

----

EDIT: Please also include a guide to fix the WebView if the package is incorrect: https://github.com/ankidroid/Anki-Android/issues/15847#issuecomment-1988304056

### PR DESCRIPTION
**Pull request by RobozinhoD**
## Purpose / Description

* Fixes #15855

## Fixes
* Fixes #15855
* Fixes #15876

## Approach

Show a dialog with a link to the wiki if the webview version is < 77

## How Has This Been Tested?

In the emulator with API 25

image

## Checklist
_Please, go through these checks before submitting the PR._

- [X] You have a descriptive commit message with a short title (first line, max 50 chars).
- [X] You have commented your code, particularly in hard-to-understand areas
- [X] You have performed a self-review of your own code
- [X] UI changes: include screenshots of all affected screens (in particular showing any new or changed strings)
- [ ] UI Changes: You have tested your change using the [Google Accessibility Scanner](https://play.google.com/store/apps/details?id=com.google.android.apps.accessibility.auditor)

### CODE REVIEW

File: AnkiDroid/src/main/res/values/03-dialogs.xml [hunk 1]
```diff
@@ -277,4 +277,6 @@ also changes the interval of the card"
     <string name="open">Open</string>
     <string name="tap_to_cloze">Single tap text to cloze</string>
 
+    <!-- Outdated Webview dialog -->
+    <string name="webview_update_message">The system Webview is outdated. Some features won\'t work correctly. Please update it.</string>
```
Comment (david-allison) [inline AnkiDroid/src/main/res/values/03-dialogs.xml:281]:
```suggestion
    <string name="webview_update_message">The system Webview is outdated. Some features won’t work correctly. Please update it.</string>
```

**Review by david-allison — CHANGES_REQUESTED** [review]
Looks good, one nitpick, cheers!

Optional: provide the user with debug info lising versions

**Comment by RobozinhoD** [comment]
image

**Comment by mikehardy** [comment]
It is not pending merge, there is an outstanding rejection that @david-allison should clear @criticalAY - please don't change that label

**Comment by neeldoshii** [comment]
Thank you for contributing. Can you provide co-author to [aayush-tripathi](https://github.com/aayush-tripathi) as credits to the original PR.

**Review by david-allison — APPROVED** [review]
Looks good, cheers!

**Comment by david-allison** [comment]
Great catch @neeldoshii

File: AnkiDroid/src/main/java/com/ichi2/anki/InitialActivity.kt [hunk 2]
```diff
@@ -133,6 +138,44 @@ object InitialActivity {
...
+            showOutdatedWebViewDialog(activity, versionCode, activity.getString(R.string.link_legacy_webview_update))
+        } else {
+            Timber.w("WebView is outdated. %s: %s", webviewPackageInfo.packageName, webviewPackageInfo.versionName)
+            showOutdatedWebViewDialog(activity, versionCode, activity.getString(R.string.link_webview_update))
+        }
+    }
+
+    private fun showOutdatedWebViewDialog(activity: AnkiActivity, installedVersion: Int, learnMoreUrl: String) {
+        AlertDialog.Builder(activity).show {
+            setMessage(activity.getString(R.string.webview_update_message, installedVersion, OLDEST_WORKING_WEBVIEW_VERSION))
```
Comment (neeldoshii) [inline AnkiDroid/src/main/java/com/ichi2/anki/InitialActivity.kt:165]: Can we use `.message`, `.positveButton` from our extension methods

**Review by neeldoshii — CHANGES_REQUESTED** [review]

File: AnkiDroid/src/main/java/com/ichi2/anki/InitialActivity.kt (hunk 2 above)
Comment (RobozinhoD) [reply AnkiDroid/src/main/java/com/ichi2/anki/InitialActivity.kt:165]: I don't know if you have read them, but it says `Allows easier transformations from [MaterialDialog] to [AlertDialog].`

Other than that, they are just platform methods duplicates, and changing them it's six of one and half a dozen of the other.

Comment (RobozinhoD) [reply AnkiDroid/src/main/java/com/ichi2/anki/InitialActivity.kt:165]: Those methods shouldn't exist now that MaterialDialog was removed

**Comment by neeldoshii** [comment]
@RobozinhoD here's an updated patch.
```diff

Subject: [PATCH] Uses AlertDialog Extension Methods
---
Index: AnkiDroid/src/main/java/com/ichi2/anki/InitialActivity.kt
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/AnkiDroid/src/main/java/com/ichi2/anki/InitialActivity.kt b/AnkiDroid/src/main/java/com/ichi2/anki/InitialActivity.kt
--- a/AnkiDroid/src/main/java/com/ichi2/anki/InitialActivity.kt	(revision 457ad443a0a66a0b1a5ab264c5d758a67dda9d68)
+++ b/AnkiDroid/src/main/java/com/ichi2/anki/InitialActivity.kt	(date 1717600569972)
@@ -38,7 +38,8 @@
 import com.ichi2.utils.Permissions
 import com.ichi2.utils.VersionUtils.pkgVersionName
 import com.ichi2.utils.getAndroidSystemWebViewPackageInfo
-import com.ichi2.utils.show
+import com.ichi2.utils.message
+import com.ichi2.utils.positiveButton
 import kotlinx.parcelize.Parcelize
 import net.ankiweb.rsdroid.BackendException
 import timber.log.Timber
@@ -161,12 +162,11 @@
     }
 
     private fun showOutdatedWebViewDialog(activity: AnkiActivity, installedVersion: Int, learnMoreUrl: String) {
-        AlertDialog.Builder(activity).show {
-            setMessage(activity.getString(R.string.webview_update_message, installedVersion, OLDEST_WORKING_WEBVIEW_VERSION))
-            setPositiveButton(R.string.scoped_storage_learn_more) { _, _ ->
+        AlertDialog.Builder(activity)
+            .message(text = activity.getString(R.string.webview_update_message, installedVersion, OLDEST_WORKING_WEBVIEW_VERSION))
+            .positiveButton(stringRes = R.string.scoped_storage_learn_more) {
                 activity.openUrl(learnMoreUrl)
-            }
-        }
+            }.show()
     }
 
     private fun getLegacyWebViewPackageInfo(packageManager: PackageManager): PackageInfo? {
     
```

**Comment by mikehardy** [comment]
I just cleaned up one ellipse on crowdin to clear out strings, and I'm running a string sync, there appears to be ongoing discussion here with patch proposal so not sure this is actually pending merge ?

**Comment by neeldoshii** [comment]
@RobozinhoD here's an updated patch.
```diff
Subject: [PATCH] Uses AlertDialog Extension Methods
---
Index: AnkiDroid/src/main/java/com/ichi2/anki/InitialActivity.kt
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/AnkiDroid/src/main/java/com/ichi2/anki/InitialActivity.kt b/AnkiDroid/src/main/java/com/ichi2/anki/InitialActivity.kt
--- a/AnkiDroid/src/main/java/com/ichi2/anki/InitialActivity.kt	(revision 457ad443a0a66a0b1a5ab264c5d758a67dda9d68)
+++ b/AnkiDroid/src/main/java/com/ichi2/anki/InitialActivity.kt	(date 1717600840786)
@@ -38,6 +38,8 @@
 import com.ichi2.utils.Permissions
 import com.ichi2.utils.VersionUtils.pkgVersionName
 import com.ichi2.utils.getAndroidSystemWebViewPackageInfo
+import com.ichi2.utils.message
+import com.ichi2.utils.positiveButton
 import com.ichi2.utils.show
 import kotlinx.parcelize.Parcelize
 import net.ankiweb.rsdroid.BackendException
@@ -162,8 +164,8 @@
 
     private fun showOutdatedWebViewDialog(activity: AnkiActivity, installedVersion: Int, learnMoreUrl: String) {
         AlertDialog.Builder(activity).show {
-            setMessage(activity.getString(R.string.webview_update_message, installedVersion, OLDEST_WORKING_WEBVIEW_VERSION))
-            setPositiveButton(R.string.scoped_storage_learn_more) { _, _ ->
+            message(text = activity.getString(R.string.webview_update_message, installedVersion, OLDEST_WORKING_WEBVIEW_VERSION))
+            positiveButton(stringRes = R.string.scoped_storage_learn_more) {
                 activity.openUrl(learnMoreUrl)
             }
         }
```

**Comment by BrayanDSO** [comment]
The defense of the PR author is adequate and the patch is just unnecessary churn

**Comment by mikehardy** [comment]
okay, just pending strings then. Those are being irritating but will go through shortly I think

**Comment by mikehardy** [comment]
strings will merge shortly, finally got them through, they're in the merge queue. I'm turning attention away so anyone else that can, throw this in the queue and do a string sync after, or if not I'll get it much later from now (and do the post-merge string sync...)

**Comment by mikehardy** [comment]
oh, actually realized that putting this in the queue is not blocked.
But! Someone will need to run a strings sync after this merges so that it's nice and clean with only the automated changes

**Pull request merged**